### PR TITLE
adding pyupgrade & autoflake

### PR DESCRIPTION
**Pull request by leos**
Adding [pyupgrade](https://github.com/asottile/pyupgrade) and [autoflake](https://github.com/PyCQA/autoflake) to the linting in `pre-commit`.

Pyupgrade cleans up a lot of syntax, some of it legacy Python 2. Because many files use `from __future__ import annotations` it also converts the typehints to a more modern 3.10 syntax using `|` instead of `Optional` and names for types like `list` or `dict` so that `List` and `Dict` no longer need to be imported.

`autoflake` then automatically cleans up the no longer used imports.

This PR is almost entirely mechanical:
```
pre-commit run pyupgrade -a
pre-commit run autoflake -a
```

There were just a few typehint issues that this revealed that `mypy` then caught that had to be fixed manually:
```
dataprofiler/labelers/data_processing.py:1540: error: Incompatible default for argument "random_state" (default has type "None", argument has type "Union[Random, int, List[Any], Tuple[Any, ...]]")
dataprofiler/labelers/data_processing.py:1882: error: Incompatible default for argument "priority_order" (default has type "None", argument has type "Union[List[Any], ndarray[Any, Any]]")
dataprofiler/labelers/data_processing.py:1883: error: Incompatible default for argument "random_state" (default has type "None", argument has type "Union[Random, int, List[Any], Tuple[Any, ...]]")
dataprofiler/labelers/data_processing.py:2089: error: Incompatible default for argument "random_state" (default has type "None", argument has type "Union[Random, int, List[Any], Tuple[Any, ...]]")
dataprofiler/labelers/base_data_labeler.py:766: error: Incompatible default for argument "labels" (default has type "None", argument has type "Union[List[Any], Dict[Any, Any]]")
Found 5 errors in 2 files (checked 39 source files)
```

### CODE REVIEW

**Comment by taylorfturner** [comment]
@leos, for clarity on expectations, I may not get to review / approve this PR today. However, I will at least read through it and provide any initial comments. Best!

**Comment by leos** [comment]
Thanks @taylorfturner ! No rush whatsover - this isn't blocking anything for me. Happy to make tweaks/updates as folks prefer.

**Comment by taylorfturner** [comment]
> Thanks @taylorfturner ! No rush whatsover - this isn't blocking anything for me. Happy to make tweaks/updates as folks prefer.

Sound good -- just for clarity for everyone I will pop a `medium priority` on this PR in the mean time. Cheers!

File: dataprofiler/labelers/classification_report_utils.py
Comment (ksneab7) [inline dataprofiler/labelers/classification_report_utils.py:328]: Is there a reason for using "format" as opposed to f strings like in above strings?

Comment (leos) [reply dataprofiler/labelers/classification_report_utils.py:328]: pyupgrade tries to clean up what it can - it's not 100% perfect. In this case it would make sense to turn the whole string into an f-string but it is figuring out that using `{0}` and `{1}` is redundant.

Happy to make a manual edit here to turn this into an f-string if you prefer.

**Comment by micdavis** [comment]
Other than Kenny's comment, LGTM. Here is a summary of changes that occur. Please call out any that I missed:

### Lowercase for certain typing such as dict and list:
`def label_mapping(self) -> Dict:` -> `def label_mapping(self) -> dict:`
`def labels(self) -> List[str]:` -> `def labels(self) -> list[str]:`

### Optional type changes:
`def preprocessor(self) -> Optional[data_processing.BaseDataPreprocessor]:` -> `def preprocessor(self) -> data_processing.BaseDataPreprocessor | None:`

### Union typing changes: 
`def set_labels(self, labels: Union[List, Dict]) -> None:` -> `def set_labels(self, labels: list | dict) -> None:`

### Remove subclass from super: 
`return super(CSVData, self)._get_data_as_records(data)` -> `return super()._get_data_as_records(data)`


### Removal of “r” from open() calls:
` with open(input_file["path"], "r") as fp:` -> `with open(input_file["path"]) as fp:`

### Removal of additional parenthesis:
```
rstr = float(
    (
        header_check_list[i].count("str")
        + header_check_list[i].count("upstr")
        + header_check_list[i].count("none")
    )
)
```
->
```
rstr = float(
    header_check_list[i].count("str")
    + header_check_list[i].count("upstr")
    + header_check_list[i].count("none")
)
```

### Formatted string change:
`expected_error = ["{}.sample_ratio must be a float.".format(optpth)]` -> `expected_error = [f"{optpth}.sample_ratio must be a float."]`

### Remove six from tests:
`six.assertCountEqual(self, ["nan"], profile.null_types)` -> `self.assertCountEqual(["nan"], profile.null_types)`

**Comment by leos** [comment]
@micdavis there's a few more
### Get rid of old style class definition
`class SpreadSheetDataMixin(object):` -> `class SpreadSheetDataMixin:`

### Remove unnecessary future imports
`from __future__ import absolute_import, print_function` -> ` `

### Explicit raw string definitions
`"header \(0 based index\)",` -> `r"header \(0 based index\)",`

### Replace undocumented assert methods
`self.assertEquals` -> `self.assertEqual`

File: dataprofiler/labelers/classification_report_utils.py
Comment (JGSweets) [reply dataprofiler/labelers/classification_report_utils.py:328]: I'm okay either way for now, but I am more a fan of being consistent.

Comment (leos) [reply dataprofiler/labelers/classification_report_utils.py]: No problem - I've fixed these. Others can be fixed opportunistically. FYI, pylint is configured to flag strings that can be f-strings - there's actually a lot of pylint warnings/suggestions that are flagged by the current config, they all just show up by default in VS Code.

**Comment by leos** [comment]
Just checking in - if folks can approve the PR if they're ok with it that would be great. Given that it touches a lot of files it risks diverging/conflicts the longer it sits unmerged.

Alternatively if folks would prefer another approach (or not adding pyupgrade) then happy to close this out.

**Comment by JGSweets** [comment]
> Just checking in - if folks can approve the PR if they're ok with it that would be great. Given that it touches a lot of files it risks diverging/conflicts the longer it sits unmerged.
> 
> Alternatively if folks would prefer another approach (or not adding pyupgrade) then happy to close this out.

I'm going to make an intent to review Friday. Are you able to update the branch in the meantime? Thanks!

**Comment by leos** [comment]
Thanks @JGSweets - yup, I'll update the branch before Friday

**Comment by taylorfturner** [comment]
@leos, by way of update, I will review this tomorrow. CC @JGSweets

File: dataprofiler/labelers/base_data_labeler.py
Comment (taylorfturner) [inline dataprofiler/labelers/base_data_labeler.py:8]: `flake8` caught this and `autoflake` modified. Still concerned with the other changes that `autoflake` is making

**Review by taylorfturner — COMMENTED** [review]
Comments and with the removal of `six` in testing suite, we would also need to remove `six>=1.15.0` from `requirements.txt`

File: dataprofiler/tests/profilers/test_base_column_profilers.py
Comment (taylorfturner) [reply dataprofiler/tests/profilers/test_base_column_profilers.py:166]: still has `import six` in the file

File: dataprofiler/tests/profilers/test_graph_profiler.py
Comment (taylorfturner) [reply dataprofiler/tests/profilers/test_graph_profiler.py:5]: @JGSweets was this an important part specifically for graph somehow, though?

File: .pre-commit-config.yaml
Comment (leos) [reply .pre-commit-config.yaml:70]: Sorry, yes, this is from autoflake - it is in their docs as the last line of their [Introduction](https://github.com/PyCQA/autoflake#introduction:
```
autoflake also removes useless pass statements by default.
```
This can be disabled trivially in the pre-commit config (`--ignore-pass-statements` or `--ignore-pass-after-docstring` as args), let me know if you'd like me to do that.

File: dataprofiler/data_readers/data_utils.py
Comment (leos) [reply dataprofiler/data_readers/data_utils.py:5]: Just to clarify, dataprofiler is long past python2 compatibility, right?

File: .pre-commit-config.yaml
Comment (taylorfturner) [reply .pre-commit-config.yaml:70]: ah there it is -- I think I'd prefer to leave pass in for now. So let's add those flags in the `pre-commit` section for `autoflake`.

File: dataprofiler/labelers/char_load_tf_model.py
Comment (leos) [reply dataprofiler/labelers/char_load_tf_model.py:199]: `pyupgrade` does it because "r" is the default mode, so while it's explicit, it's redundant.

File: dataprofiler/data_readers/data_utils.py
Comment (taylorfturner) [reply dataprofiler/data_readers/data_utils.py:5]: Yes, the minimum we focus on supporting right now is 3.8 I believe

File: dataprofiler/labelers/classification_report_utils.py
Comment (leos) [reply dataprofiler/labelers/classification_report_utils.py:329]: The proposed change will add a lot of spaces to the final string (ie. between "of" and "target_names") - is that what we want here?

File: dataprofiler/tests/profilers/test_base_column_profilers.py
Comment (leos) [reply dataprofiler/tests/profilers/test_base_column_profilers.py:166]: `autoflake` by default only catches unused imports that come from standardlib. Out of the box cleaning up all seemingly unused imports can be risky because they may be relied on as side effects. `autoflake` can be used with the following settings to adjust the default behavior:
```
 --imports IMPORTS     by default, only unused standard library imports are removed; specify a comma-separated list of additional modules/packages
  --expand-star-imports
                        expand wildcard star imports with undefined names; this only triggers if there is only one star import in the file; this is skipped if there are any uses of `__all__` or `del` in the file
  --remove-all-unused-imports
                        remove all unused imports (not just those from the standard library)
```

File: dataprofiler/labelers/char_load_tf_model.py
Comment (taylorfturner) [reply dataprofiler/labelers/char_load_tf_model.py:199]: yeah that's fair, I'm good with this diff

**Comment by leos** [comment]
> we would also need to remove `six>=1.15.0` from `requirements.txt`

Yup - I was planning to do that as a followup PR. Remove all imports of `six` and then remove it from the requirements, running `pipdeptree` to remove any transitive deps it left behind as well.

File: .pre-commit-config.yaml
Comment (leos) [reply .pre-commit-config.yaml:70]: No problem - `--ignore-pass-statements` or `--ignore-pass-after-docstring` or both?

File: dataprofiler/labelers/classification_report_utils.py
Comment (taylorfturner) [reply dataprofiler/labelers/classification_report_utils.py:329]: no, not looking for a bunch of spaces ... I've never seen it render with all a bunch of spaces when formatting it like that, but I totally could be mistaken in this situation.

File: dataprofiler/tests/profilers/test_base_column_profilers.py
Comment (taylorfturner) [reply dataprofiler/tests/profilers/test_base_column_profilers.py:166]: i see

File: dataprofiler/data_readers/data_utils.py
Comment (leos) [reply dataprofiler/data_readers/data_utils.py:5]: Cool, if that's the case, we can bump up the syntax version in the `pre-commit` config to `--py38-plus`. Might make sense to do that as a separate PR to minimize the noise.

File: dataprofiler/labelers/classification_report_utils.py
Comment (leos) [reply dataprofiler/labelers/classification_report_utils.py:329]:
```
In [4]: a = """asdf
   ...:        cats
   ...:        dogs"""

In [5]: a
Out[5]: 'asdf\n       cats\n       dogs'
```

Multiline strings don't do anything smart with indenting unfortunately.

Comment (leos) [reply dataprofiler/labelers/classification_report_utils.py:329]: Or specifically in a function call:
```
In [11]: print("""asdf
    ...:        cats
    ...:        dogs""")
asdf
       cats
       dogs

In [12]: print("asdf "
    ...:       "cats "
    ...:       "dogs")
asdf cats dogs
```

File: .pre-commit-config.yaml
Comment (taylorfturner) [reply .pre-commit-config.yaml:70]: For now, maybe let's just do `--ignore-pass-statements`. Thx

File: dataprofiler/data_readers/data_utils.py
Comment (taylorfturner) [reply dataprofiler/data_readers/data_utils.py:5]: yeah agree

File: dataprofiler/labelers/classification_report_utils.py
Comment (taylorfturner) [reply dataprofiler/labelers/classification_report_utils.py:329]: I was mistaken. I'll close this comment and keep as original in your PR

File: .pre-commit-config.yaml
Comment (JGSweets) [reply .pre-commit-config.yaml:63]: Should this be `py38-plus`? We can always update in a follow-up PR so we don't have to re-review everything again.

File: dataprofiler/tests/profilers/test_graph_profiler.py
Comment (JGSweets) [reply dataprofiler/tests/profilers/test_graph_profiler.py:5]: @taylorfturner My assumption is this was left over from a previous test looking through visual output reports at one point for verification.

I think it is fine to remove since we aren't using it.

**Review by JGSweets — DISMISSED** [review]

**Comment by JGSweets** [comment]
@leos  Thank you for working on this and giving us time to review it in full. Definitely a worthwhile bit of cleanup!

File: .pre-commit-config.yaml
Comment (JGSweets) [reply .pre-commit-config.yaml:70]: I'm good either way. I approved the PR for now even though this is not in yet.

I'll re-review quickly after this change has been made.

File: dataprofiler/data_readers/avro_data.py
Comment (JGSweets) [reply dataprofiler/data_readers/avro_data.py:61]: To be fair, this code here could use a clean-up as it is not ideal to have a function such that we need to pass mypy.

File: .pre-commit-config.yaml
Comment (leos) [reply .pre-commit-config.yaml:63]: Yup, will do that in a followup PR

Comment (leos) [reply .pre-commit-config.yaml]: Actually, just did it here, it only had one tiny change - https://github.com/capitalone/DataProfiler/pull/734/commits/d6afd70347406a028effabf82438a1b3b7ee392b - moving the import for `Protocol` from `typing_extensions` to `typing.

File: dataprofiler/labelers/regex_model.py
Comment (leos) [reply dataprofiler/labelers/regex_model.py:248]: This would be a functional change - the formatting syntax (ie. `:d` is the same in f-strings as it is with `format()`.

If y'all would like me to do that, happy to.

**Comment by leos** [comment]
Addressed all PR feedback:
* Restored the unnecessary `pass` statements and disabled `autoflake` from removing them
* Moved `pyupgrade` to `py38-plus`

https://github.com/capitalone/DataProfiler/pull/734/files/e7e259443304917f0c8d6debe3a10abac5a35468..d6afd70347406a028effabf82438a1b3b7ee392b

Let me know if there's anything else before this can land.

File: dataprofiler/labelers/regex_model.py
Comment (JGSweets) [reply dataprofiler/labelers/regex_model.py:248]: Correct, `:d` is already formatting.

Let's leave it as is

**Review by JGSweets — APPROVED** [review]

**Review by taylorfturner — APPROVED** [review]
LGTM! Thanks for the contribution and bearing with us on the timeline. Cheers!

**Pull request merged**
